File: go/progress.go
package pintui

import (
	"fmt"
	"os"
	"time"

	"github.com/briandowns/spinner"
	"github.com/fatih/color"
	"github.com/schollz/progressbar/v3"
)

// Spinner creates and starts a spinner for indeterminate progress.
//
// The spinner animates while work is in progress, showing the user
// that the application hasn't frozen.
//
// Example:
//
//	s := pintui.Spinner("Loading configuration...")
//	// ... do work ...
//	s.Success("Configuration loaded")
func Spinner(msg string) *SpinnerHandle {
	s := spinner.New(spinner.CharSets[14], 80*time.Millisecond) // CharSet 14 is ⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏
	s.Suffix = " " + msg
	s.Color("cyan")
	s.Start()
	return &SpinnerHandle{spinner: s, msg: msg}
}

// SpinnerHandle wraps a spinner and provides finish methods.
type SpinnerHandle struct {
	spinner *spinner.Spinner
	msg     string
}

// Success stops the spinner and shows a success message.
//
// Example:
//
//	s := pintui.Spinner("Connecting...")
//	// ... do work ...
//	s.Success("Connected")
//	// Output: ✓ Connected
func (s *SpinnerHandle) Success(msg string) {
	s.spinner.Stop()
	fmt.Printf("\r%s %s\n", color.GreenString("✓"), msg)
}

// Error stops the spinner and shows an error message.
//
// Example:
//
//	s := pintui.Spinner("Connecting...")
//	// ... connection fails ...
//	s.Error("Connection failed")
//	// Output: ✗ Connection failed
func (s *SpinnerHandle) Error(msg string) {
	s.spinner.Stop()
	fmt.Fprintf(os.Stderr, "\r%s %s\n", color.RedString("✗"), msg)
}

// Warn stops the spinner and shows a warning message.
//
// Example:
//
//	s := pintui.Spinner("Checking...")
//	// ... partial success ...
//	s.Warn("Completed with warnings")
//	// Output: ⚠ Completed with warnings
func (s *SpinnerHandle) Warn(msg string) {
	s.spinner.Stop()
	fmt.Printf("\r%s %s\n", color.YellowString("⚠"), msg)
}

// Clear stops the spinner and clears it from display.
//
// Use when you want to remove the spinner without leaving a final message.
func (s *SpinnerHandle) Clear() {
	s.spinner.Stop()
	fmt.Print("\r\033[K") // Clear line
}

// UpdateMessage updates the spinner's message while it's running.
func (s *SpinnerHandle) UpdateMessage(msg string) {
	s.msg = msg
	s.spinner.Suffix = " " + msg
}

// Bar creates a progress bar for determinate progress.
//
// Use when you know the total amount of work to be done.
//
// Example:
//
//	bar := pintui.Bar(100, "Downloading")
//	for i := 0; i < 100; i++ {
//	    // ... do work ...
//	    bar.Add(1)
//	}
//	bar.Success("Downloaded")
func Bar(total int64, prefix string) *BarHandle {
	bar := progressbar.NewOptions64(total,
		progressbar.OptionSetDescription(prefix),
		progressbar.OptionSetWidth(40),
		progressbar.OptionEnableColorCodes(true),
		progressbar.OptionSetTheme(progressbar.Theme{
			Saucer:        "[cyan]━[reset]",
			SaucerHead:    "[cyan]╸[reset]",
			SaucerPadding: "[blue]─[reset]",
			BarStart:      "[",
			BarEnd:        "]",
		}),
		progressbar.OptionShowCount(),
		progressbar.OptionSpinnerType(14),
	)
	return &BarHandle{bar: bar}
}

// BarHandle wraps a progress bar and provides helper methods.
type BarHandle struct {
	bar *progressbar.ProgressBar
}

// Add increments the progress bar by the given amount.
func (b *BarHandle) Add(n int) {
	b.bar.Add(n)
}

// Add64 increments the progress bar by the given int64 amount.
func (b *BarHandle) Add64(n int64) {
	b.bar.Add64(n)
}

// Set sets the progress bar to a specific value.
func (b *BarHandle) Set(n int) {
	b.bar.Set(n)
}

// Set64 sets the progress bar to a specific int64 value.
func (b *BarHandle) Set64(n int64) {
	b.bar.Set64(n)
}

// Success finishes the progress bar with a success message.
func (b *BarHandle) Success(msg string) {
	b.bar.Finish()
	fmt.Printf("\r%s %s\n", color.GreenString("✓"), msg)
}

// Error finishes the progress bar with an error message.
func (b *BarHandle) Error(msg string) {
	b.bar.Finish()
	fmt.Fprintf(os.Stderr, "\r%s %s\n", color.RedString("✗"), msg)
}

// Clear finishes and clears the progress bar.
func (b *BarHandle) Clear() {
	b.bar.Clear()
}

// StageProgress tracks progress through a sequence of named stages.
//
// Example:
//
//	stages := pintui.NewStageProgress(3)
//
//	s1 := stages.Next("Downloading")
//	// ... download ...
//	s1.Success("Downloaded")
//
//	s2 := stages.Next("Extracting")
//	// ... extract ...
//	s2.Success("Extracted")
//
//	stages.Skip("Installing") // Skip if not needed
type StageProgress struct {
	current int
	total   int
}

// NewStageProgress creates a new stage progress tracker.
func NewStageProgress(total int) *StageProgress {
	return &StageProgress{current: 0, total: total}
}

// Next starts the next stage and returns a spinner for it.
//
// The spinner shows [current/total] before the stage name.
func (sp *StageProgress) Next(name string) *SpinnerHandle {
	sp.current++
	s := spinner.New(spinner.CharSets[14], 80*time.Millisecond)
	s.Prefix = fmt.Sprintf("[%d/%d] ", sp.current, sp.total)
	s.Suffix = " " + name
	s.Color("cyan")
	s.Start()
	return &SpinnerHandle{spinner: s, msg: name}
}

// Skip marks a stage as skipped.
func (sp *StageProgress) Skip(name string) {
	sp.current++
	fmt.Printf("  %s [%d/%d] %s (skipped)\n",
		color.New(color.Faint).Sprint("○"),
		sp.current, sp.total, name)
}

// Current returns the current stage number (1-indexed).
func (sp *StageProgress) Current() int {
	return sp.current
}

// Total returns the total number of stages.
func (sp *StageProgress) Total() int {
	return sp.total
}

// IsComplete returns true if all stages are complete.
func (sp *StageProgress) IsComplete() bool {
	return sp.current >= sp.total
}

File: rust/src/progress.rs
//! Progress indicators for long-running operations.
//!
//! This module provides spinners and progress bars for giving users
//! feedback during operations that take time.
//!
//! ## Spinner (Indeterminate Progress)
//!
//! Use when you don't know how long an operation will take:
//!
//! ```no_run
//! use pintui::progress;
//!
//! let spinner = progress::spinner("Connecting to server...");
//! // ... do work ...
//! progress::finish_success(&spinner, "Connected");
//! ```
//!
//! ## Progress Bar (Determinate Progress)
//!
//! Use when you know the total amount of work:
//!
//! ```no_run
//! use pintui::progress;
//!
//! let bar = progress::bar(100, "Downloading");
//! for i in 0..100 {
//!     // ... do work ...
//!     bar.inc(1);
//! }
//! progress::finish_success(&bar, "Downloaded");
//! ```
//!
//! ## Multi-Stage Progress
//!
//! Use for operations with multiple distinct phases:
//!
//! ```no_run
//! use pintui::progress::StageProgress;
//!
//! let mut stages = StageProgress::new(3);
//!
//! let s1 = stages.next("Fetching");
//! // ... fetch ...
//! pintui::progress::finish_success(&s1, "Fetched");
//!
//! let s2 = stages.next("Building");
//! // ... build ...
//! pintui::progress::finish_success(&s2, "Built");
//! ```

use colored::Colorize;
use indicatif::{ProgressBar, ProgressStyle};
use std::time::Duration;

/// Create a spinner for indeterminate progress.
///
/// The spinner animates while work is in progress, showing the user
/// that the application hasn't frozen.
///
/// # Example
///
/// ```no_run
/// let spinner = pintui::progress::spinner("Loading configuration...");
/// // ... do work ...
/// pintui::progress::finish_success(&spinner, "Configuration loaded");
/// ```
pub fn spinner(msg: &str) -> ProgressBar {
    let pb = ProgressBar::new_spinner();
    pb.set_style(
        ProgressStyle::default_spinner()
            .tick_chars("⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏")
            .template("{spinner:.cyan} {msg}")
            .unwrap(),
    );
    pb.set_message(msg.to_string());
    pb.enable_steady_tick(Duration::from_millis(80));
    pb
}

/// Create a progress bar for determinate progress.
///
/// Use when you know the total amount of work to be done.
///
/// # Arguments
///
/// * `len` - Total number of units of work
/// * `prefix` - Label shown before the progress bar
///
/// # Example
///
/// ```no_run
/// let bar = pintui::progress::bar(100, "Processing");
/// for _ in 0..100 {
///     // ... do work ...
///     bar.inc(1);
/// }
/// pintui::progress::finish_success(&bar, "Processed 100 items");
/// ```
pub fn bar(len: u64, prefix: &str) -> ProgressBar {
    let pb = ProgressBar::new(len);
    pb.set_style(
        ProgressStyle::default_bar()
            .template(
                "{prefix:.bold.dim} {spinner:.green} [{bar:40.cyan/blue}] {pos}/{len} {wide_msg}",
            )
            .unwrap()
            .progress_chars("━━╸"),
    );
    pb.set_prefix(prefix.to_string());
    pb.enable_steady_tick(Duration::from_millis(100));
    pb
}

/// Finish a progress indicator with a success message.
///
/// Replaces the spinner/bar with a green checkmark and message.
///
/// # Example
///
/// ```no_run
/// let spinner = pintui::progress::spinner("Working...");
/// // ... do work ...
/// pintui::progress::finish_success(&spinner, "Work complete");
/// // Output: ✓ Work complete
/// ```
pub fn finish_success(pb: &ProgressBar, msg: &str) {
    pb.set_style(ProgressStyle::default_spinner().template("{msg}").unwrap());
    pb.finish_with_message(format!("{} {}", "✓".green(), msg));
}

/// Finish a progress indicator with an error message.
///
/// Replaces the spinner/bar with a red cross and message.
///
/// # Example
///
/// ```no_run
/// let spinner = pintui::progress::spinner("Connecting...");
/// // ... connection fails ...
/// pintui::progress::finish_error(&spinner, "Connection failed");
/// // Output: ✗ Connection failed
/// ```
pub fn finish_error(pb: &ProgressBar, msg: &str) {
    pb.set_style(ProgressStyle::default_spinner().template("{msg}").unwrap());
    pb.finish_with_message(format!("{} {}", "✗".red(), msg));
}

/// Finish a progress indicator with a warning message.
///
/// Replaces the spinner/bar with a yellow warning sign and message.
///
/// # Example
///
/// ```no_run
/// let spinner = pintui::progress::spinner("Checking...");
/// // ... partial success ...
/// pintui::progress::finish_warn(&spinner, "Completed with warnings");
/// // Output: ⚠ Completed with warnings
/// ```
pub fn finish_warn(pb: &ProgressBar, msg: &str) {
    pb.set_style(ProgressStyle::default_spinner().template("{msg}").unwrap());
    pb.finish_with_message(format!("{} {}", "⚠".yellow(), msg));
}

/// Finish a progress indicator and clear it from display.
///
/// Use when you want to remove the progress indicator without
/// leaving a final message.
pub fn finish_clear(pb: &ProgressBar) {
    pb.finish_and_clear();
}

/// Multi-stage progress tracker.
///
/// Tracks progress through a sequence of named stages, showing
/// `[current/total]` indicators.
///
/// # Example
///
/// ```no_run
/// use pintui::progress::{StageProgress, finish_success};
///
/// let mut stages = StageProgress::new(3);
///
/// let s1 = stages.next("Downloading");
/// // ... download ...
/// finish_success(&s1, "Downloaded");
///
/// let s2 = stages.next("Extracting");
/// // ... extract ...
/// finish_success(&s2, "Extracted");
///
/// let s3 = stages.next("Installing");
/// // ... install ...
/// finish_success(&s3, "Installed");
/// ```
pub struct StageProgress {
    current: usize,
    total: usize,
}

impl StageProgress {
    /// Create a new stage progress tracker.
    ///
    /// # Arguments
    ///
    /// * `total` - Total number of stages
    #[must_use]
    pub const fn new(total: usize) -> Self {
        Self { current: 0, total }
    }

    /// Start the next stage and return a spinner for it.
    ///
    /// The spinner shows `[current/total]` before the stage name.
    pub fn next(&mut self, name: &str) -> ProgressBar {
        self.current += 1;
        let pb = ProgressBar::new_spinner();
        pb.set_style(
            ProgressStyle::default_spinner()
                .tick_chars("⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏")
                .template(&format!(
                    "{{spinner:.cyan}} [{}/{}] {{msg}}",
                    self.current, self.total
                ))
                .unwrap(),
        );
        pb.set_message(name.to_string());
        pb.enable_steady_tick(Duration::from_millis(80));
        pb
    }

    /// Skip a stage, printing a skip indicator.
    ///
    /// Increments the stage counter but doesn't create a spinner.
    pub fn skip(&mut self, name: &str) {
        self.current += 1;
        println!(
            "  {} [{}/{}] {} (skipped)",
            "○".dimmed(),
            self.current,
            self.total,
            name
        );
    }

    /// Get the current stage number (1-indexed).
    #[must_use]
    pub const fn current(&self) -> usize {
        self.current
    }

    /// Get the total number of stages.
    #[must_use]
    pub const fn total(&self) -> usize {
        self.total
    }

    /// Check if all stages are complete.
    #[must_use]
    pub const fn is_complete(&self) -> bool {
        self.current >= self.total
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_spinner_creation() {
        let pb = spinner("Test");
        pb.finish_and_clear();
    }

    #[test]
    fn test_bar_creation() {
        let pb = bar(100, "Test");
        pb.finish_and_clear();
    }

    #[test]
    fn test_finish_functions() {
        let pb = spinner("Test");
        finish_success(&pb, "Done");

        let pb = spinner("Test");
        finish_error(&pb, "Failed");

        let pb = spinner("Test");
        finish_warn(&pb, "Warning");

        let pb = spinner("Test");
        finish_clear(&pb);
    }

    #[test]
    fn test_stage_progress() {
        let mut stages = StageProgress::new(3);

        assert_eq!(stages.current(), 0);
        assert_eq!(stages.total(), 3);
        assert!(!stages.is_complete());

        let s1 = stages.next("Stage 1");
        assert_eq!(stages.current(), 1);
        s1.finish_and_clear();

        let s2 = stages.next("Stage 2");
        assert_eq!(stages.current(), 2);
        s2.finish_and_clear();

        stages.skip("Stage 3");
        assert_eq!(stages.current(), 3);
        assert!(stages.is_complete());
    }

    #[test]
    fn test_empty_stage_progress() {
        let stages = StageProgress::new(0);
        assert!(stages.is_complete());
    }
}
